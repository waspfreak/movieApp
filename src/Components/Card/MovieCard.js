import React, {useState} from 'react';
import PropTypes from 'prop-types';
import ProgressBar from '../ProgressBar/ProgressBar';
import moment from 'moment';

// Components
import Modal from '../Modal/Modal';
import MovieDetails from '../MovieDetail/MovieDetails';

// Custom Hook
import {useMovieFetch} from '../../hooks/useMovieFetch';
// Styled Components
import {StyledMovieCard} from './style';

// Images
import NoImage from '../../assets/no_image.jpg';

// Environment Variables
import {IMAGE_BASE_URL, POSTER_SIZE} from '../../api/API';

export const MovieCard = ({
	image,
	movieId,
	clickable,
	originalTitle,
	releaseDate,
	voteAverage,
	overview,
}) => {
	const [isModal, setModal] = useState(false);

	const [movie, error] = useMovieFetch(movieId);

	if (error) return <div>Something went wrong ...</div>;

	return (
		<>
			<StyledMovieCard onClick={() => setModal(true)}>
				<>
					{clickable ? (
						<>
							<Modal isVisible={isModal} onClose={() => setModal(false)}>
								<img src={image} alt={originalTitle} />
								<MovieDetails
									modalOverview={overview}
									detailTitle={originalTitle}
									movie={movie}
									posterPath={
										movie.poster_path
											? `${IMAGE_BASE_URL}${POSTER_SIZE}${image}`
											: NoImage
									}
									voteAverage={voteAverage}
								/>
							</Modal>

							<img className='clickable' src={image} alt={originalTitle} />
							<div className='content'>
								<div className='score'>
									<ProgressBar
										progress={parseFloat(voteAverage).toFixed(1) * 10}
										size={50}
										strokeWidth={4}
										circleOneStroke='#3c0a0adb'
										circleTwoStroke='red'
									/>
								</div>

								<h2 className='content-title'>{originalTitle}</h2>
								<p className='content-data'>
									{moment(releaseDate).format('LL')}
								</p>
							</div>
						</>
					) : (
						<img src={image} alt={originalTitle} />
					)}
				</>
			</StyledMovieCard>
		</>
	);
};

MovieCard.propTypes = {
	image: PropTypes.string,
	movieId: PropTypes.number,
	clickable: PropTypes.bool,
	originalTitle: PropTypes.string,
	releaseDate: PropTypes.string,
};

export default MovieCard;
