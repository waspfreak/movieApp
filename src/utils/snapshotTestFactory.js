import {cleanup, render} from '@testing-library/react';
import React from 'react';

/**
 *
 * Sets up snapshot tests in a standard way and avoids repeating boiler plate
 */
export function snapshotTestFactory(Component: any, properties: any) {
	/**
	 * Re-usable it block
	 */
	const itBlock = (props: any) => {
		it('should match snapshot', () => {
			const {container} = render(<Component {...props} />);
			expect(container).toMatchSnapshot();
		});
	};

	/**
	 * Re-usable describe block
	 */
	describe(`${Component.name}`, () => {
		/**
		 * Clean up any mock functions that might be in props
		 */
		afterEach(() => {
			jest.clearAllMocks();
			cleanup();
		});

		/**
		 * If its an array of props then iterate them into seperate it blocks
		 */
		if (Array.isArray(properties)) {
			properties.forEach((props) => {
				itBlock(props);
			});
		} else {
			itBlock(properties);
		}
	});
}
