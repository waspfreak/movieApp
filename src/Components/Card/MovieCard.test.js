import {snapshotTestFactory} from '../../utils/snapshotTestFactory';
//
import {MovieCard} from './MovieCard';
const props: MovieCard[] = [
	{
		image: 'http://image.tmdb.org/t/p/w500/lPsD10PP4rgUGiGR4CCXA6iY0QQ.jpg',
		movieId: 16,
		clickable: true,
		originalTitle: 'Movie Original title',
		releaseDate: '2001, January 1',
		voteAverage: '55',
	},
];
snapshotTestFactory(MovieCard, props);
