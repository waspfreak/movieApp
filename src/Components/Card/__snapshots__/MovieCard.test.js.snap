// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard should match snapshot 1`] = `
.c1 {
  background: #111;
  border-radius: 50%;
  padding: 2px;
  display: inline-block;
}

.c2 {
  display: block;
  margin: 0px auto;
  max-width: 100%;
}

.c3 {
  fill: none;
}

.c4 {
  fill: none;
}

.c5 {
  font-family: 'Titillium Web',sans-serif;
  font-size: 14px;
  text-anchor: middle;
  fill: white;
  font-weight: bold;
  color: white;
}

.c0 {
  position: unset;
  font-family: 'Titillium Web',sans-serif;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-content: flex-start;
  -ms-flex-line-pack: start;
  align-content: flex-start;
  -webkit-transition: -webkit-transform 1.45s;
  -webkit-transition: transform 1.45s;
  transition: transform 1.45s;
  max-width: 250px;
  width: 100%;
  margin-inline: 10px;
  margin-bottom: 20px;
  cursor: pointer;
  opacity: 0.9;
}

.c0:hover {
  -webkit-transform: scale(0.95);
  -ms-transform: scale(0.95);
  transform: scale(0.95);
  opacity: 1;
}

.c0 .content {
  width: 100%;
  position: relative;
  white-space: normal;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: flex-start;
  -ms-flex-line-pack: start;
  align-content: flex-start;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow-wrap: break-word;
  background-color: #2b2b2a;
  padding: 10px;
  min-height: 100px;
}

.c0 .content .content-title {
  overflow-wrap: break-word;
  font-weight: 500;
  width: 200px;
  font-size: 1.2em;
  margin-bottom: 10px;
}

.c0 .content .content-data {
  font-size: 16px;
  color: #948c8c;
}

.c0 .score {
  position: absolute;
  top: -28%;
  right: 1%;
  z-index: 9999;
}

.c0 img {
  width: 100%;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  object-fit: cover;
  border: none;
}

@media screen and (max-width:1369) {
  .c0 {
    max-width: 243px;
    width: 100%;
  }
}

@media screen and (max-width:1023px) {
  .c0 {
    max-width: 243px;
    width: 100%;
  }
}

@media screen and (max-width:770px) {
  .c0 {
    max-width: 215px;
    width: 100%;
  }
}

@media screen and (max-width:600px) {
  .c0 {
    max-width: 100%;
    width: 100%;
  }
}

@media screen and (max-width:1023px) {
  .c0 {
    max-width: 243px;
    width: 100%;
  }
}

@media screen and (max-width:1370) {
  .c0 {
    max-width: 243px;
    width: 100%;
  }
}

@media screen and (max-width:770px) {
  .c0 {
    max-width: 320px;
    width: 100%;
  }
}

@media screen and (max-width:600px) {
  .c0 {
    max-width: 100%;
    width: 100%;
  }
}

<div>
  <div
    class="c0"
  >
    <img
      alt="Movie Original title"
      class="clickable"
      src="http://image.tmdb.org/t/p/w500/lPsD10PP4rgUGiGR4CCXA6iY0QQ.jpg"
    />
    <div
      class="content"
    >
      <div
        class="score"
      >
        <div
          class="c1"
        >
          <svg
            class="c2"
            height="50"
            width="50"
          >
            <circle
              class="c3"
              cx="25"
              cy="25"
              r="23"
              stroke="#3c0a0adb"
              stroke-width="4"
            />
            <circle
              class="c4"
              cx="25"
              cy="25"
              r="23"
              stroke="red"
              stroke-dasharray="144.51326206513048"
              stroke-dashoffset="65.03096792930872"
              stroke-width="4"
              style="transition: stroke-dashoffset 850ms ease-in-out;"
            />
            <text
              class="c5"
              x="25"
              y="25"
            >
              55
              %
            </text>
          </svg>
        </div>
      </div>
      <h2
        class="content-title"
      >
        Movie Original title
      </h2>
      <p
        class="content-data"
      >
        <time
          datetime="978307200000"
        >
          Mon Jan 01 2001 00:00:00 GMT+0000
        </time>
      </p>
    </div>
  </div>
</div>
`;
